File: src/routes/health.ts
import type { FastifyPluginAsync } from 'fastify';
import { pingDb } from '../db/index.js';
import { registry } from '../metrics.js';
import { redis } from '../redis.js';
import { limiter } from '../riot/limiter.js';
import { config } from '../config.js';
import {
  HealthResponse,
  LimitsResponse,
  ReadyResponse,
  ScopeParam,
  localErrors,
} from './schemas.js';
import { Type } from '@sinclair/typebox';

/** §6.2 / FR-12 — liveness and Prometheus. Both are public (no consumer key). */
const healthRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get(
    '/healthz',
    {
      config: { public: true },
      schema: {
        tags: ['ops'],
        summary: 'Liveness',
        description:
          'Answers as long as the process is up. It touches neither Redis nor Postgres, so a ' +
          '200 here says only that the event loop is turning — use `/readyz` to decide whether ' +
          'to send traffic.',
        response: { 200: HealthResponse },
      },
    },
    async () => {
      return { ok: true };
    },
  );

  // Deeper check for load balancers and the compose healthcheck.
  fastify.get(
    '/readyz',
    {
      config: { public: true },
      schema: {
        tags: ['ops'],
        summary: 'Readiness',
        description:
          'Pings Redis and Postgres. **The 503 is the interesting response**: it carries the ' +
          'same body as the 200, so the `redis` and `postgres` booleans name which dependency ' +
          'is the reason the service is not ready. Used by the compose healthcheck and by any ' +
          'load balancer that should stop routing rather than serve errors.',
        response: { 200: ReadyResponse, 503: ReadyResponse },
      },
    },
    async (_request, reply) => {
      const [redisOk, dbOk] = await Promise.all([
        redis
          .ping()
          .then(() => true)
          .catch(() => false),
        pingDb(),
      ]);
      const ok = redisOk && dbOk;
      if (!ok) reply.code(503);
      return { ok, redis: redisOk, postgres: dbOk, keyScope: config.KEY_SCOPE };
    },
  );

  fastify.get(
    '/metrics',
    {
      config: { public: true },
      schema: {
        tags: ['ops'],
        summary: 'Prometheus metrics',
        description:
          'The §13 metric set in Prometheus text exposition format — not JSON. In the shipped ' +
          'Caddy configuration this path is reachable only from private ranges, so a public ' +
          'deployment will see a 403 from the proxy in front rather than a response from here.',
        // Declared as `content` rather than a bare schema: a response schema
        // keyed by status alone would hand this route to fast-json-stringify,
        // which would JSON-encode the exposition text.
        response: {
          200: {
            description: 'Prometheus text exposition format',
            content: { 'text/plain': { schema: Type.String() } },
          },
        },
      },
    },
    async (_request, reply) => {
      reply.header('Content-Type', registry.contentType);
      return registry.metrics();
    },
  );

  // Operator visibility into current bucket usage (§9) without a Grafana trip.
  fastify.get(
    '/v1/admin/limits/:scope',
    {
      config: { scope: 'admin' },
      schema: {
        tags: ['admin'],
        summary: 'Rate-limit bucket usage',
        description:
          'Current token usage for one Riot rate-limit bucket, and how long any 429-induced ' +
          'freeze on it has left to run. The scope is a host, not a game region: platform ' +
          'endpoints bucket by platform (`euw1`), account-v1 and match-v5 by region (`europe`).',
        params: Type.Object({ scope: ScopeParam }),
        response: { 200: LimitsResponse, ...localErrors },
      },
    },
    async (request) => {
      const { scope } = request.params as { scope: string };
      const [usage, frozenMs] = await Promise.all([limiter.usage(scope), limiter.isFrozen(scope)]);
      return { scope, usage, frozenMs };
    },
  );
};

export default healthRoutes;

File: src/routes/health.rs
//! `/healthz` and `/readyz`. Both are public: no key, no quota (v1 `routes/health.ts`).
//! `/metrics` is served by `telemetry::metrics_router`.

use std::time::Duration;

use axum::Json;
use axum::extract::State;
use axum::http::StatusCode;
use serde::Serialize;
use utoipa::ToSchema;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;

use crate::app::AppState;
use crate::cache::keys::KeyScope;
use crate::db::DbError;

/// A readiness probe that cannot get a write slot this quickly counts as not ready.
const READY_TIMEOUT: Duration = Duration::from_secs(2);

pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(healthz))
        .routes(routes!(readyz))
}

#[derive(Debug, Serialize, ToSchema)]
pub struct Health {
    pub ok: bool,
}

/// Liveness: answers while the process is up. Touches nothing.
#[utoipa::path(
    get,
    path = "/healthz",
    tag = "ops",
    summary = "Liveness",
    description = "Answers as long as the process is up. It touches nothing, so a 200 says only that \
                   the process is serving; use `/readyz` to decide whether to send traffic.",
    security(()),
    responses((status = 200, description = "Up", body = Health)),
)]
async fn healthz() -> Json<Health> {
    Json(Health { ok: true })
}

/// v1 returned `{ok, redis, postgres, keyScope}`. v2 has one store, so `sqlite` replaces
/// the two backend booleans, and `limiter` says the checkpoint was restored
/// (design/07). `keyScope` is v1's (ADR-011).
#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct Ready {
    pub ok: bool,
    pub sqlite: bool,
    pub limiter: bool,
    pub key_scope: String,
}

/// Readiness: the writer thread is alive and can take SQLite's write lock, and the
/// limiter checkpoint has been restored.
/// 503 carries the same body, so it names what is not ready (v1 behaviour).
#[utoipa::path(
    get,
    path = "/readyz",
    tag = "ops",
    summary = "Readiness",
    description = "SQLite can take a write and the limiter checkpoint is restored. **The 503 carries the \
                   same body as the 200**, so the booleans name what is not ready.",
    security(()),
    responses(
        (status = 200, description = "Ready", body = Ready),
        (status = 503, description = "Not ready", body = Ready),
    ),
)]
async fn readyz(State(state): State<AppState>) -> (StatusCode, Json<Ready>) {
    let probe = state.db.write(|c| {
        c.execute_batch("BEGIN IMMEDIATE; ROLLBACK;")
            .map_err(DbError::from)
    });
    let sqlite = matches!(tokio::time::timeout(READY_TIMEOUT, probe).await, Ok(Ok(())));
    let limiter = state.limiter_restored.load(std::sync::atomic::Ordering::Acquire);
    let ok = sqlite && limiter;
    let status = if ok {
        StatusCode::OK
    } else {
        StatusCode::SERVICE_UNAVAILABLE
    };
    (
        status,
        Json(Ready {
            ok,
            sqlite,
            limiter,
            key_scope: KeyScope::from_key(&state.config.riot_api_key).to_string(),
        }),
    )
}
